src/vulkan/device.cc

```diff
@@ -403,121 +402,134 @@ Result Device::Initialize(
+                                       required_features)) {
+    return Result(
+        "Vulkan: Device::Initialize given physical device does not support "
+        "required features");
   }
 
-  VkPhysicalDeviceFeatures available_vulkan_features =
-      VkPhysicalDeviceFeatures();
-  if (use_physical_device_features_2) {
-    available_vulkan_features = available_features2.features;
-
-    VkPhysicalDeviceVariablePointerFeaturesKHR* var_ptrs = nullptr;
-    VkPhysicalDeviceFloat16Int8FeaturesKHR* float16_ptrs = nullptr;
-    VkPhysicalDevice8BitStorageFeaturesKHR* storage8_ptrs = nullptr;
-    VkPhysicalDevice16BitStorageFeaturesKHR* storage16_ptrs = nullptr;
-    VkPhysicalDeviceSubgroupSizeControlFeaturesEXT*
-        subgroup_size_control_features = nullptr;
-    void* ptr = available_features2.pNext;
-    while (ptr != nullptr) {
-      BaseOutStructure* s = static_cast<BaseOutStructure*>(ptr);
-      if (s->sType ==
-          VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_VARIABLE_POINTER_FEATURES_KHR) {
-        var_ptrs =
-            static_cast<VkPhysicalDeviceVariablePointerFeaturesKHR*>(ptr);
-      } else if (s->sType ==
-                 VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_FLOAT16_INT8_FEATURES_KHR) {
-        float16_ptrs =
-            static_cast<VkPhysicalDeviceFloat16Int8FeaturesKHR*>(ptr);
-      } else if (s->sType ==
-                 VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_8BIT_STORAGE_FEATURES_KHR) {
-        storage8_ptrs =
-            static_cast<VkPhysicalDevice8BitStorageFeaturesKHR*>(ptr);
-      } else if (s->sType ==
-                 VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_16BIT_STORAGE_FEATURES_KHR) {
-        storage16_ptrs =
-            static_cast<VkPhysicalDevice16BitStorageFeaturesKHR*>(ptr);
-      } else if (
-          s->sType ==
-          VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_SUBGROUP_SIZE_CONTROL_FEATURES_EXT) {  // NOLINT(whitespace/line_length)
-        subgroup_size_control_features =
-            static_cast<VkPhysicalDeviceSubgroupSizeControlFeaturesEXT*>(ptr);
-      }
-      ptr = s->pNext;
+  // Search for additional features in case they are found in pNext field of
+  // available_features2.
+  VkPhysicalDeviceVariablePointerFeaturesKHR* var_ptrs = nullptr;
+  VkPhysicalDeviceFloat16Int8FeaturesKHR* float16_ptrs = nullptr;
+  VkPhysicalDevice8BitStorageFeaturesKHR* storage8_ptrs = nullptr;
+  VkPhysicalDevice16BitStorageFeaturesKHR* storage16_ptrs = nullptr;
+  VkPhysicalDeviceVulkan11Features* vulkan11_ptrs = nullptr;
+  VkPhysicalDeviceVulkan12Features* vulkan12_ptrs = nullptr;
+  VkPhysicalDeviceSubgroupSizeControlFeaturesEXT*
+      subgroup_size_control_features = nullptr;
+  void* ptr = available_features2.pNext;
+  while (ptr != nullptr) {
+    BaseOutStructure* s = static_cast<BaseOutStructure*>(ptr);
+    if (s->sType ==
+        VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_VARIABLE_POINTER_FEATURES_KHR) {
+      var_ptrs = static_cast<VkPhysicalDeviceVariablePointerFeaturesKHR*>(ptr);
```

Comment:
I know this is not due to your change, but can we make this a switch?

Reply:
sure